// @wix/cli/build/chunk-QKXNIB4D.js
import { createRequire as _createRequire } from 'node:module';
const require = _createRequire(import.meta.url);
import {
  DeploymentPipelinesStatus
} from "./chunk-6XRFJZTQ.js";
import {
  z
} from "./chunk-GNB7T7J7.js";
import {
  init_esm_shims
} from "./chunk-PYIAC2GK.js";

// ../cli-site-old/src/schemas.ts
init_esm_shims();
var deployAppResponseSchema = z.object({
  deploymentInfo: z.object({
    deploymentId: z.string(),
    deploymentUrl: z.string().url(),
    deploymentShortUrl: z.string().nullish(),
    deployedRevision: z.string(),
    isPublishPipelineDeployment: z.boolean()
  })
});
var resolveNpmDependenciesSchema = z.object({
  jobId: z.string().uuid()
});
var npmDependenciesLockFileResponseSchema = z.object({
  result: z.object({
    npmDependenciesLockFile: z.string().url()
  })
});
var getDeploymentResponseSchema = z.object({
  deployment: z.object({
    deploymentPipelinesStatus: z.enum([
      DeploymentPipelinesStatus.IN_PROGRESS,
      DeploymentPipelinesStatus.SUCCESS,
      DeploymentPipelinesStatus.ERROR
    ])
  })
});

export {
  deployAppResponseSchema,
  resolveNpmDependenciesSchema,
  npmDependenciesLockFileResponseSchema,
  getDeploymentResponseSchema
};
