// @wix/cli/build/LogoutCommand-AQ4FFG6S.js
import { createRequire as _createRequire } from 'node:module';
const require = _createRequire(import.meta.url);
import {
  logout
} from "./chunk-O25O35J2.js";
import "./chunk-7VND5WMC.js";
import "./chunk-WN3D6AOY.js";
import {
  Box_default,
  ConfirmInput,
  Spinner,
  Text,
  Trans,
  require_react,
  useAsyncCallback,
  useTranslation
} from "./chunk-OH3G6JMX.js";
import "./chunk-W423UTD3.js";
import "./chunk-2K5KKUL3.js";
import "./chunk-YU54OBFT.js";
import "./chunk-XMTHMR2B.js";
import "./chunk-F6DVF5WQ.js";
import "./chunk-GNB7T7J7.js";
import {
  require_lib
} from "./chunk-CVJG3VZM.js";
import {
  __toESM,
  init_esm_shims
} from "./chunk-PYIAC2GK.js";

// src/components/LogoutCommand/index.ts
init_esm_shims();

// src/components/LogoutCommand/LogoutCommand.tsx
init_esm_shims();
var import_react = __toESM(require_react(), 1);
var import_variant = __toESM(require_lib(), 1);
var LogoutCommand = ({ userInfo }) => {
  const { t } = useTranslation();
  const logoutAction = useAsyncCallback(logout);
  if (userInfo == null) {
    return /* @__PURE__ */ import_react.default.createElement(Trans, { i18nKey: "logout_command.not_logged_in" });
  }
  return /* @__PURE__ */ import_react.default.createElement(import_react.default.Fragment, null, /* @__PURE__ */ import_react.default.createElement(Box_default, { marginBottom: 1 }, /* @__PURE__ */ import_react.default.createElement(Trans, { i18nKey: "logout_command.logout_warning" })), /* @__PURE__ */ import_react.default.createElement(
    ConfirmInput,
    {
      label: t("logout_command.logout_confirm"),
      initialValue: true,
      onSubmit: (confirmed) => {
        if (confirmed) {
          void logoutAction.execute();
        }
      }
    }
  ), (0, import_variant.match)(logoutAction.status, {
    NotRequested: () => null,
    Error: () => null,
    Loading: () => /* @__PURE__ */ import_react.default.createElement(Spinner, null),
    Success: () => /* @__PURE__ */ import_react.default.createElement(Text, null, t("logout_command.logout_success"))
  }));
};
export {
  LogoutCommand
};
