// @wix/cli/build/PreviewCommand-PF63VX7E.js
import { createRequire as _createRequire } from 'node:module';
const require = _createRequire(import.meta.url);
import {
  DeploymentStatus,
  useCreatePreview
} from "./chunk-YNXR6XGC.js";
import "./chunk-LZJPV4CP.js";
import "./chunk-6XRFJZTQ.js";
import "./chunk-KQMZOXQO.js";
import "./chunk-HGBMRF3C.js";
import "./chunk-E56WW2OA.js";
import "./chunk-SNWTG76C.js";
import "./chunk-6LI4WGAW.js";
import "./chunk-DDC4C6V6.js";
import "./chunk-BALHGZ6G.js";
import "./chunk-Q6F4OGQT.js";
import "./chunk-O25O35J2.js";
import "./chunk-7VND5WMC.js";
import "./chunk-WN3D6AOY.js";
import {
  Alert,
  Badge,
  Link,
  Spinner,
  Trans,
  require_react,
  useAsync
} from "./chunk-OH3G6JMX.js";
import "./chunk-W423UTD3.js";
import "./chunk-2K5KKUL3.js";
import "./chunk-YU54OBFT.js";
import "./chunk-XMTHMR2B.js";
import "./chunk-2CPXK3HU.js";
import "./chunk-F6DVF5WQ.js";
import "./chunk-GNB7T7J7.js";
import {
  require_lib
} from "./chunk-CVJG3VZM.js";
import {
  __toESM,
  init_esm_shims
} from "./chunk-PYIAC2GK.js";

// ../cli-site/src/components/PreviewCommand/PreviewCommand.tsx
init_esm_shims();
var import_react = __toESM(require_react(), 1);
var import_variant = __toESM(require_lib(), 1);
var PreviewCommand = () => {
  const createPreview = useCreatePreview();
  const { status } = useAsync(async () => {
    const { previewUrl, deploymentStatus } = await createPreview();
    if ((0, import_variant.isType)(deploymentStatus, DeploymentStatus.Error)) {
      throw deploymentStatus.deploymentError;
    }
    return previewUrl;
  }, []);
  return (0, import_variant.match)(status, {
    Error: () => null,
    Loading: () => /* @__PURE__ */ import_react.default.createElement(Spinner, { text: /* @__PURE__ */ import_react.default.createElement(Trans, { i18nKey: "preview_command.loading" }) }),
    Success: ({ result }) => /* @__PURE__ */ import_react.default.createElement(Alert, { type: "success" }, /* @__PURE__ */ import_react.default.createElement(Badge, { skin: "success" }, /* @__PURE__ */ import_react.default.createElement(Trans, { i18nKey: "general.success" })), " ", /* @__PURE__ */ import_react.default.createElement(
      Trans,
      {
        i18nKey: "preview_command.success",
        components: [/* @__PURE__ */ import_react.default.createElement(Link, { url: result })]
      }
    ))
  });
};
export {
  PreviewCommand
};
