// @wix/cli/build/PublishCommand-QIUB5P27.js
import { createRequire as _createRequire } from 'node:module';
const require = _createRequire(import.meta.url);
import {
  DeploymentStatus,
  useCreatePreview,
  useDocumentManagementClient
} from "./chunk-YNXR6XGC.js";
import "./chunk-LZJPV4CP.js";
import "./chunk-6XRFJZTQ.js";
import "./chunk-KQMZOXQO.js";
import "./chunk-HGBMRF3C.js";
import "./chunk-E56WW2OA.js";
import "./chunk-SNWTG76C.js";
import "./chunk-6LI4WGAW.js";
import "./chunk-DDC4C6V6.js";
import "./chunk-BALHGZ6G.js";
import "./chunk-Q6F4OGQT.js";
import "./chunk-O25O35J2.js";
import {
  ErrorViewer
} from "./chunk-7VND5WMC.js";
import "./chunk-WN3D6AOY.js";
import {
  Alert,
  Badge,
  Box_default,
  ConfirmInput,
  Link,
  Spinner,
  Trans,
  require_react,
  useAsync,
  useAsyncCallback,
  useTranslation
} from "./chunk-OH3G6JMX.js";
import "./chunk-W423UTD3.js";
import "./chunk-2K5KKUL3.js";
import "./chunk-YU54OBFT.js";
import "./chunk-XMTHMR2B.js";
import "./chunk-2CPXK3HU.js";
import "./chunk-F6DVF5WQ.js";
import "./chunk-GNB7T7J7.js";
import {
  require_lib
} from "./chunk-CVJG3VZM.js";
import {
  __toESM,
  init_esm_shims
} from "./chunk-PYIAC2GK.js";

// ../cli-site/src/components/PublishCommand/PublishCommand.tsx
init_esm_shims();
var import_react = __toESM(require_react(), 1);
var import_variant = __toESM(require_lib(), 1);
var PublishCommand = ({ approvePreview, force }) => {
  const { t } = useTranslation();
  const createPreview = useCreatePreview();
  const { publishDeployment } = useDocumentManagementClient();
  const publishDeploymentAction = useAsyncCallback(
    async (_, deploymentId) => publishDeployment(deploymentId)
  );
  const { status } = useAsync(async () => {
    const preview = await createPreview();
    const preApprovePreview = force || ((0, import_variant.isType)(preview.deploymentStatus, DeploymentStatus.Success) ? approvePreview : false);
    if (preApprovePreview) {
      void publishDeploymentAction.execute(preview.deploymentId);
    }
    return {
      ...preview,
      preApprovePreview
    };
  }, []);
  return (0, import_variant.match)(status, {
    Error: () => null,
    Loading: () => /* @__PURE__ */ import_react.default.createElement(Spinner, { text: /* @__PURE__ */ import_react.default.createElement(Trans, { i18nKey: "publish_command.preview.loading" }) }),
    Success: ({ result }) => {
      return /* @__PURE__ */ import_react.default.createElement(import_react.default.Fragment, null, (0, import_variant.match)(result.deploymentStatus, {
        Success: () => /* @__PURE__ */ import_react.default.createElement(Alert, { type: "success" }, /* @__PURE__ */ import_react.default.createElement(
          Trans,
          {
            i18nKey: "publish_command.preview.success",
            components: [/* @__PURE__ */ import_react.default.createElement(Link, { url: result.previewUrl })]
          }
        )),
        Error: ({ deploymentError }) => /* @__PURE__ */ import_react.default.createElement(import_react.default.Fragment, null, /* @__PURE__ */ import_react.default.createElement(ErrorViewer, { error: deploymentError }), /* @__PURE__ */ import_react.default.createElement(Box_default, { marginTop: 1 }, /* @__PURE__ */ import_react.default.createElement(Alert, { type: "warning" }, /* @__PURE__ */ import_react.default.createElement(Trans, { i18nKey: "publish_command.preview.warning" }))))
      }), !result.preApprovePreview && /* @__PURE__ */ import_react.default.createElement(Box_default, { marginTop: 1 }, /* @__PURE__ */ import_react.default.createElement(
        ConfirmInput,
        {
          label: t("publish_command.confirm_publish"),
          onSubmit: (value) => {
            if (value) {
              void publishDeploymentAction.execute(result.deploymentId);
            }
          }
        }
      )), (0, import_variant.match)(publishDeploymentAction.status, {
        NotRequested: () => null,
        Error: () => null,
        Loading: () => /* @__PURE__ */ import_react.default.createElement(Box_default, { marginTop: 1 }, /* @__PURE__ */ import_react.default.createElement(
          Spinner,
          {
            text: /* @__PURE__ */ import_react.default.createElement(Trans, { i18nKey: "publish_command.publish.loading" })
          }
        )),
        Success: () => /* @__PURE__ */ import_react.default.createElement(Box_default, { marginTop: 1 }, /* @__PURE__ */ import_react.default.createElement(Alert, { type: "success" }, /* @__PURE__ */ import_react.default.createElement(Badge, { skin: "success" }, /* @__PURE__ */ import_react.default.createElement(Trans, { i18nKey: "general.success" })), " ", /* @__PURE__ */ import_react.default.createElement(
          Trans,
          {
            i18nKey: "publish_command.publish.success",
            components: [/* @__PURE__ */ import_react.default.createElement(Link, { url: result.siteUrl })]
          }
        )))
      }));
    }
  });
};
export {
  PublishCommand
};
